File: apps/ios/Sources/Utils/CommandQueue.swift
//
// SmartBLE - Command Queue (iOS)
//

import Foundation
import CoreBluetooth

/// 指令状态
enum CommandStatus: String {
    case pending = "pending"
    case sending = "sending"
    case success = "success"
    case failed = "failed"
}

/// 指令队列项
struct CommandItem: Identifiable {
    let id: String
    let deviceId: String
    let serviceUuid: String
    let characteristicUuid: String
    let data: Data
    let withoutResponse: Bool
    let displayHex: String
    
    var status: CommandStatus = .pending
    var error: String? = nil
    var sentAt: Date? = nil
}

/// 指令发送回调 (Send Closure)
typealias CommandSender = (CommandItem) async throws -> Void

/// BLE 指令队列管理器
@MainActor
class CommandQueue: ObservableObject {
    @Published private(set) var queue: [CommandItem] = []
    @Published private(set) var history: [CommandItem] = []
    
    @Published var isRunning: Bool = false
    @Published var isPaused: Bool = false
    @Published var isLooping: Bool = false
    
    @Published var targetLoopCount: Int = 0
    @Published var currentLoop: Int = 0
    
    var intervalMs: Int = 50
    
    private let sender: CommandSender
    private var loopTemplate: [CommandItem]? = nil
    
    // Callbacks
    var onCommandStart: ((CommandItem) -> Void)?
    var onCommandComplete: ((CommandItem) -> Void)?
    var onCommandError: ((CommandItem) -> Void)?
    var onQueueEmpty: (() -> Void)?
    
    init(intervalMs: Int = 50, sender: @escaping CommandSender) {
        self.intervalMs = intervalMs
        self.sender = sender
    }
    
    var pendingCount: Int {
        return queue.count
    }
    
    func enqueue(_ command: CommandItem) {
        queue.append(command)
        startProcessing()
    }
    
    func enqueueBatch(_ commands: [CommandItem]) {
        queue.append(contentsOf: commands)
        startProcessing()
    }
    
    func startLoop(_ commands: [CommandItem], loopCount: Int = 0) {
        stopLoop()
        isLooping = true
        targetLoopCount = loopCount
        currentLoop = 0
        loopTemplate = commands
        enqueueNextLoop()
    }
    
    func stopLoop() {
        isLooping = false
        loopTemplate = nil
        targetLoopCount = 0
        currentLoop = 0
    }
    
    func pause() {
        isPaused = true
    }
    
    func resume() {
        isPaused = false
        startProcessing()
    }
    
    func clear() {
        queue.removeAll()
        stopLoop()
        isRunning = false
        isPaused = false
    }
    
    func clearHistory() {
        history.removeAll()
    }
    
    private func startProcessing() {
        if isRunning || isPaused || queue.isEmpty {
            return
        }
        isRunning = true
        Task {
            await processNext()
        }
    }
    
    private func processNext() async {
        if isPaused || queue.isEmpty {
            isRunning = false
            if queue.isEmpty && !isLooping {
                onQueueEmpty?()
            }
            if queue.isEmpty && isLooping {
                enqueueNextLoop()
                if !queue.isEmpty {
                    await processNext()
                } else {
                    isRunning = false
                }
            }
            return
        }
        
        var command = queue.removeFirst()
        command.status = .sending
        command.sentAt = Date()
        self.onCommandStart?(command)
        
        do {
            try await sender(command)
            command.status = .success
            self.onCommandComplete?(command)
        } catch {
            command.status = .failed
            command.error = error.localizedDescription
            self.onCommandError?(command)
        }
        
        history.append(command)
        
        if !queue.isEmpty || isLooping {
            try? await Task.sleep(nanoseconds: UInt64(intervalMs) * 1_000_000)
            await processNext()
        } else {
            isRunning = false
            self.onQueueEmpty?()
        }
    }
    
    private func enqueueNextLoop() {
        guard isLooping, let template = loopTemplate else { return }
        
        if targetLoopCount > 0 && currentLoop >= targetLoopCount {
            isLooping = false
            loopTemplate = nil
            onQueueEmpty?()
            return
        }
        
        currentLoop += 1
        
        let newCommands = template.map { item in
            CommandItem(
                id: "\(item.id)_loop\(currentLoop)",
                deviceId: item.deviceId,
                serviceUuid: item.serviceUuid,
                characteristicUuid: item.characteristicUuid,
                data: item.data,
                withoutResponse: item.withoutResponse,
                displayHex: item.displayHex
            )
        }
        
        queue.append(contentsOf: newCommands)
    }
}

File: core/ble-core/utils/command-queue.ts
import { logger } from './logger';

export type CommandStatus = 'pending' | 'sending' | 'success' | 'failed';

export interface CommandItem {
    id: string;
    deviceId: string;
    serviceUuid: string;
    characteristicUuid: string;
    data: Uint8Array;
    withoutResponse?: boolean;
    displayHex: string;
    status: CommandStatus;
    error?: string;
    sentAt?: Date;
}

export type CommandSender = (command: CommandItem) => Promise<void>;

export interface CommandQueueOptions {
    intervalMs?: number;
    onCommandStart?: (command: CommandItem) => void;
    onCommandComplete?: (command: CommandItem) => void;
    onCommandError?: (command: CommandItem) => void;
    onQueueEmpty?: () => void;
    onQueueStateChanged?: () => void;
}

export class CommandQueue {
    private queue: CommandItem[] = [];
    private history: CommandItem[] = [];
    
    public intervalMs: number = 50;
    private _isRunning = false;
    private _isPaused = false;
    
    private _isLooping = false;
    private _targetLoopCount = 0;
    private _currentLoop = 0;
    private _loopTemplate: CommandItem[] | null = null;
    
    private sender: CommandSender;
    private options: CommandQueueOptions;
    
    constructor(sender: CommandSender, options: CommandQueueOptions = {}) {
        this.sender = sender;
        this.options = options;
        if (options.intervalMs !== undefined) {
            this.intervalMs = options.intervalMs;
        }
    }
    
    get pendingCount(): number { return this.queue.length; }
    get isRunning(): boolean { return this._isRunning; }
    get isPaused(): boolean { return this._isPaused; }
    get isLooping(): boolean { return this._isLooping; }
    get currentLoop(): number { return this._currentLoop; }
    get targetLoopCount(): number { return this._targetLoopCount; }
    get getQueue(): CommandItem[] { return [...this.queue]; }
    get getHistory(): CommandItem[] { return [...this.history]; }
    
    enqueue(command: CommandItem): void {
        this.queue.push(command);
        this.options.onQueueStateChanged?.();
        this.startProcessing();
    }
    
    enqueueBatch(commands: CommandItem[]): void {
        this.queue.push(...commands);
        this.options.onQueueStateChanged?.();
        this.startProcessing();
    }
    
    startLoop(commands: CommandItem[], loopCount: number = 0): void {
        this.stopLoop();
        this._isLooping = true;
        this._targetLoopCount = loopCount;
        this._currentLoop = 0;
        this._loopTemplate = commands;
        this.enqueueNextLoop();
        this.options.onQueueStateChanged?.();
    }
    
    stopLoop(): void {
        this._isLooping = false;
        this._loopTemplate = null;
        this._targetLoopCount = 0;
        this._currentLoop = 0;
        this.options.onQueueStateChanged?.();
    }
    
    pause(): void {
        this._isPaused = true;
        this.options.onQueueStateChanged?.();
    }
    
    resume(): void {
        this._isPaused = false;
        this.options.onQueueStateChanged?.();
        this.startProcessing();
    }
    
    clear(): void {
        this.queue = [];
        this.stopLoop();
        this._isRunning = false;
        this._isPaused = false;
        this.options.onQueueStateChanged?.();
    }
    
    clearHistory(): void {
        this.history = [];
    }
    
    private startProcessing(): void {
        if (this._isRunning || this._isPaused || this.queue.length === 0) {
            return;
        }
        this._isRunning = true;
        this.processNext().catch(e => logger.error(`Queue error: ${e}`));
    }
    
    private async processNext(): Promise<void> {
        if (this._isPaused || this.queue.length === 0) {
            this._isRunning = false;
            if (this.queue.length === 0 && !this._isLooping) {
                this.options.onQueueEmpty?.();
            }
            if (this.queue.length === 0 && this._isLooping) {
                this.enqueueNextLoop();
                if (this.queue.length > 0) {
                    await this.processNext();
                } else {
                    this._isRunning = false;
                }
            }
            return;
        }
        
        const command = this.queue.shift()!;
        command.status = 'sending';
        command.sentAt = new Date();
        this.options.onCommandStart?.(command);
        this.options.onQueueStateChanged?.();
        
        try {
            await this.sender(command);
            command.status = 'success';
            this.options.onCommandComplete?.(command);
        } catch (e) {
            command.status = 'failed';
            command.error = e instanceof Error ? e.message : String(e);
            this.options.onCommandError?.(command);
            logger.error(`指令发送失败: ${e}`, command.deviceId);
        }
        
        this.history.push(command);
        this.options.onQueueStateChanged?.();
        
        if (this.queue.length > 0 || this._isLooping) {
            await this.sleep(this.intervalMs);
            await this.processNext();
        } else {
            this._isRunning = false;
            this.options.onQueueEmpty?.();
            this.options.onQueueStateChanged?.();
        }
    }
    
    private enqueueNextLoop(): void {
        if (!this._isLooping || !this._loopTemplate) return;
        
        if (this._targetLoopCount > 0 && this._currentLoop >= this._targetLoopCount) {
            this._isLooping = false;
            this._loopTemplate = null;
            this.options.onQueueEmpty?.();
            this.options.onQueueStateChanged?.();
            return;
        }
        
        this._currentLoop++;
        
        const newCommands = this._loopTemplate.map(item => ({
            id: `${item.id}_loop${this._currentLoop}`,
            deviceId: item.deviceId,
            serviceUuid: item.serviceUuid,
            characteristicUuid: item.characteristicUuid,
            data: new Uint8Array(item.data),
            withoutResponse: item.withoutResponse,
            displayHex: item.displayHex,
            status: 'pending' as CommandStatus
        }));
        
        this.queue.push(...newCommands);
    }
    
    private sleep(ms: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}
